### PR TITLE
Redux

### PR DESCRIPTION
**Pull request by Vadimchesh**

### CODE REVIEW

File: src/components/AddTask/AddTask.jsx [hunk 1]
```diff
@@ -1,13 +1,33 @@
-import React from "react";
+import React, { useState } from "react";
+import { useDispatch } from "react-redux";
 import styles from "./style.module.scss";
 
-function AddTask() {
+const AddTask = () => {
+  const [text, setText] = useState("");
```
Comment (i-barysiuk) [inline src/components/AddTask/AddTask.jsx]: do we really need this state in component in this case?

File: src/components/BottomControls/BottomControls.jsx [hunk 2]
```diff
@@ -1,15 +1,53 @@
 import React from "react";
-import FilterStatus from "../FilterStatus/FilterStatus";
+import { useSelector, useDispatch } from "react-redux";
+import { StatusFilters } from "../../redux/reducerFilter";
 import styles from "./style.module.scss";
 
-function BottomControls() {
+const RemainingTodos = ({ count }) => {
```
Comment (i-barysiuk) [inline src/components/BottomControls/BottomControls.jsx]: use only one component in file. If you want to divide it^create a separated file for this

File: src/components/ListTask/ListTask.jsx [hunk 3]
```diff
@@ -1,28 +1,18 @@
...
-          ></Task>
-        );
-      })}
-    </div>
-  );
-}
+const selectTodoIds = (state) => state.todos.map((todo) => todo.id);
+
+const ListTask = () => {
+  const todoIds = useSelector(selectTodoIds, shallowEqual);
```
Comment (i-barysiuk) [inline src/components/ListTask/ListTask.jsx]: why we use shallowEqual?

File: src/components/Task/Task.jsx [hunk 4]
```diff
@@ -1,25 +1,47 @@
...
 import styles from "./style.module.scss";
 import cn from "classnames";
+import { useSelector, useDispatch } from "react-redux";
+
+const selectTodoById = (state, todoId) => {
+  return state.todos.find((todo) => todo.id === todoId);
+};
+
+const TodoListItem = ({ id }) => {
+  const todo = useSelector((state) => selectTodoById(state, id));
```
Comment (i-barysiuk) [inline src/components/Task/Task.jsx]: it is not good idea. You can get content in props. In such way you will use store only in upper component and reduce render from n+1 to 1 time

File: src/redux/reducerTodo.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  return maxId + 1;
+}
+
+export default function todosReducer(state = initialState, action) {
+  switch (action.type) {
+    case "todos/todoAdded": {
+      return [
+        ...state,
+        {
+          id: nextTodoId(state),
```
Comment (i-barysiuk) [inline src/redux/reducerTodo.js]: can we use UUID here?

**Review by i-barysiuk — CHANGES_REQUESTED** [review]

File: src/components/ListTask/ListTask.jsx (hunk 3 above)
Comment (Vadimchesh) [reply src/components/ListTask/ListTask.jsx]: If we toggle a todo item, the list of IDs will be considered the same, and <ListTask> won't have to re-render. The one <Task> will get an updated todo object and re-render, but all the rest of them will still have the existing todo object and not have to re-render at all.

File: src/components/AddTask/AddTask.jsx (hunk 1 above)
Comment (Vadimchesh) [reply src/components/AddTask/AddTask.jsx]: Why not? 
We'll need to keep track of the data the user is typing into the "Add Todo" input box.

File: src/redux/reducerTodo.js (hunk 5 above)
Comment (Vadimchesh) [reply src/redux/reducerTodo.js]: good idea, I have implemented this.

File: src/components/BottomControls/BottomControls.jsx (hunk 2 above)
Comment (Vadimchesh) [reply src/components/BottomControls/BottomControls.jsx]: OK,  I have implemented this

File: .eslintcache [hunk 6]
```diff
@@ -1 +1 @@
...
...s/todo-app/todo-app/src/index.js",[],"/Users/vadimlindo/Desktop/projects/todo-app/todo-app/src/reportWebVitals.js",[],"/Users/vadimlindo/Desktop/projects/todo-app/todo-app/src/redux/store.js",[],"/Users/vadimlindo/Desktop/projects/todo-app/todo-app/src/App.jsx",[],"/Users/vadimlindo/Desktop/projects/todo-app/todo-app/src/redux/reducer.js",[],"/Users/vadimlindo/Desktop/projects/todo-app/todo-app/src/components/AddTask/AddTask.js",[],"/Users/vadimlindo/Desktop/projects/todo-app/todo-app/src/components/ListTask/ListTask.js",[],"/Users/vadimlindo/Desktop/projects/todo-app/todo-app/src/components/Footer/Footer.jsx",[],"/Users/vadimlindo/Desktop/projects/todo-app/todo-app/src/redux/reducerFilter.js",[],"/Users/vadimlindo/Desktop/projects/todo-app/todo-app/src/redux/reducerTodo.js",[],"/Users/vadimlindo/Desktop/projects/todo-app/todo-app/src/components/Task/Task.js",[],"/Users/vadimlindo/Desktop/projects/todo-app/todo-app/src/components/BottomControls/BottomControls.js",[],"/Users/vadimlindo/Desktop/projects/todo-app/todo-app/src/components/RemainingTodos/RemainingTodos.js",[],"/Users/vadimlindo/Desktop/projects/todo-app/todo-app/src/components/BottomControls/BottomContolAllClear.jsx",[]]
```
Comment (i-barysiuk) [inline .eslintcache]: remove this file from PR and add to gitignore

File: .env [hunk 7]
```diff
@@ -0,0 +1 @@
+SKIP_PREFLIGHT_CHECK=true
```
Comment (i-barysiuk) [inline .env:1]: dont upload .env file to GIT

File: config/setupTest.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
+import Enzyme, { shallow, render, mount } from 'enzyme';
+// import Adapter from "enzyme-adapter-react-16";
```
Comment (i-barysiuk) [inline config/setupTest.js]: remove comments if dont't use it

File: config/setupTest.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
+Enzyme.configure({ adapter: new Adapter() });
+import toJson from 'enzyme-to-json';
+
+// React 16 Enzyme adapter
```
Comment (i-barysiuk) [inline config/setupTest.js]: remove comments

File: cypress/fixtures/users.json [hunk 10]
```diff
@@ -0,0 +1,232 @@
+[
```
Comment (i-barysiuk) [inline cypress/fixtures/users.json]: are you really use it ? :)

File: cypress/integration/input-form.spec.js [hunk 11]
```diff
@@ -0,0 +1,60 @@
+const { useContext } = require('react');
+
+describe('Input form', () => {
+  beforeEach(() => {
+    cy.visit('/');
+  });
+
+  describe('App initialization', () => {
+    it('Loads todos on page load', () => {
+      cy.get('[data-cy=textTask]').should('have.length', 1);
```
Comment (i-barysiuk) [inline cypress/integration/input-form.spec.js]: after app init we user shouldn't have any tasks

File: cypress/integration/input-form.spec.js [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    });
+  });
+
+  describe('Input form', () => {
+    it('focuses inpit on load', () => {
+      cy.focused().should('have.class', 'new-todo');
```
Comment (i-barysiuk) [inline cypress/integration/input-form.spec.js]: 1) we should use dynamic class from scss module 
2) classname very unstable var in development. For testing we use special data atributes ([data-cy=textTask] or [data-testId=textTask])

File: cypress/integration/input-form.spec.js [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    });
+    it('accept input"', () => {
+      const typedText = 'Buy a vodka';
+
+      cy.get('.new-todo').type(typedText).should('have.value', typedText);
+    });
+
+    context('Form submission', () => {
+      beforeEach(() => {
+        cy.server();
```
Comment (i-barysiuk) [inline cypress/integration/input-form.spec.js]: we don't implement any server side
are you sure to use it?

File: cypress/integration/input-form.spec.js [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+      });
+
+      it('Adds a new todo on submit', () => {
+        const itemText = 'Buy eggs';
+        cy.route('POST', '/api/todos', {
+          name: itemText,
+          id: 1,
+          isComplete: false,
+        });
```
Comment (i-barysiuk) [inline cypress/integration/input-form.spec.js]: our app dont have any routing

File: cypress/integration/input-form.spec.js [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+      });
+    });
+    describe('List items', () => {
+      it('properly displays completed items', () => {
+        cy.get('[data-cy=cy-task] ')
+          .filter('.completed')
+          .should('have.length', 1)
+          .and('contain', 'Eggs')
+          .find('.toggle')
+          .should('be.checked');
```
Comment (i-barysiuk) [inline cypress/integration/input-form.spec.js]: can you test all 3 filter state in this case

File: cypress/integration/input-form.spec.js [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+      });
+
+      it('Shows remaining todos in the footer', () => {
+        cy.get('[data-cy=cy-count]').should('contain', 3);
```
Comment (i-barysiuk) [inline cypress/integration/input-form.spec.js]: can you test changing of this value?

File: src/components/ListTask/ListTask.test.js [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+  const StatusFilters = {
+    All: 'Test',
+    Active: 'Test1',
+    Completed: 'Test2',
+  };
+
+  const initialState = {
+    status: StatusFilters.Test,
+  };
+  let store = mockStore(initialState);
```
Comment (i-barysiuk) [inline src/components/ListTask/ListTask.test.js]: why we use it here?

File: src/components/RemainingTodos/ReminingTodo.test.js [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+
+describe('Remmaining todo coun:', () => {
+  it('should render one label', () => {
+    const label = wrapper.find('label');
+    expect(label.length).toBe(1);
+  });
+  it('should render count', () => {
+    const label = wrapper.find('label');
+    expect(label.text()).toBe(`items left`);
+  });
```
Comment (i-barysiuk) [inline src/components/RemainingTodos/ReminingTodo.test.js]: why we dont test a number from store?

File: src/components/Task/Task.test.js [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+  });
+
+  it('snapshot', () => {
+    expect(wrapper).toMatchSnapshot();
+  });
+
+  it('called toggle fn', () => {
+    wrapper.find('.toggle').simulate('change');
+    expect(useDispatch).toBeCalled();
+    expect(mockDispatch).toBeCalled();
```
Comment (i-barysiuk) [inline src/components/Task/Task.test.js:31]: which one dispatch are we call here?

File: src/components/Task/Task.test.js [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+  });
+
+  it('called onDelete fn', () => {
+    wrapper.find('#delete').simulate('click');
+    expect(useDispatch).toBeCalled();
+    expect(mockDispatch).toBeCalled();
```
Comment (i-barysiuk) [inline src/components/Task/Task.test.js:36]: which one dispatch are we call here?

File: src/redux/reducerTodo.js [hunk 21]
```diff
@@ -0,0 +1,39 @@
+import { v4 } from 'uuid';
+
+const initialState = [{ id: 0, text: 'Task Godel', completed: true }];
```
Comment (i-barysiuk) [inline src/redux/reducerTodo.js]: we should have zero task from start

File: src/redux/reducerTodo.test.js [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+      payload: 'Run the tests',
+    });
+    expect(typeof reducer[0].id).toBe('string');
+    expect(reducer[0].text).toBe('Run the tests');
+    expect(reducer[0].completed).toBe(false);
+  });
+  it('should handle todos/todoToggle', () => {
+    const reducer = todosReducer(undefined, {
+      type: 'todos/todoToggle',
+      payload: '',
```
Comment (i-barysiuk) [inline src/redux/reducerTodo.test.js]: maybe we should pass the id of task?

**Review by i-barysiuk — CHANGES_REQUESTED** [review]

**Pull request merged**